File: backend/customer-service/src/main/java/com/insurance/customerservice/service/CustomerService.java
package com.insurance.customerservice.service;

import com.insurance.customerservice.model.Customer;
import com.insurance.customerservice.repository.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class CustomerService {
    @Autowired
    private CustomerRepository customerRepository;

    public List<Customer> getAllCustomers() {
        return customerRepository.findAll();
    }

    public Optional<Customer> getCustomerById(Long id) {
        return customerRepository.findById(id);
    }

    public Customer createCustomer(Customer customer) {
        return customerRepository.save(customer);
    }

    public Customer updateCustomer(Long id, Customer customerDetails) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Customer not found with id: " + id));
        
        customer.setFirstName(customerDetails.getFirstName());
        customer.setLastName(customerDetails.getLastName());
        customer.setEmail(customerDetails.getEmail());
        customer.setAddress(customerDetails.getAddress());
        customer.setPhone(customerDetails.getPhone());
        
        return customerRepository.save(customer);
    }

    public void deleteCustomer(Long id) {
        Customer customer = customerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Customer not found with id: " + id));
        
        customerRepository.delete(customer);
    }

    public List<Customer> searchCustomersByLastName(String lastName) {
        return customerRepository.findByLastNameContainingIgnoreCase(lastName);
    }

    public List<Customer> searchCustomersByEmail(String email) {
        return customerRepository.findByEmailContainingIgnoreCase(email);
    }
}

File: frontend/src/services/customerService.ts
import api from "./api"
import type { Customer } from "../types/Customer"

const CUSTOMERS_URL = "/customers"

const getAllCustomers = async (): Promise<Customer[]> => {
  const response = await api.get<Customer[]>(CUSTOMERS_URL)
  return response.data
}

const getCustomerById = async (id: number): Promise<Customer> => {
  const response = await api.get<Customer>(`${CUSTOMERS_URL}/${id}`)
  return response.data
}

const createCustomer = async (customer: Customer): Promise<Customer> => {
  const response = await api.post<Customer>(CUSTOMERS_URL, customer)
  return response.data
}

const updateCustomer = async (id: number, customer: Customer): Promise<Customer> => {
  const response = await api.put<Customer>(`${CUSTOMERS_URL}/${id}`, customer)
  return response.data
}

const deleteCustomer = async (id: number): Promise<void> => {
  await api.delete(`${CUSTOMERS_URL}/${id}`)
}

const searchCustomers = async (searchTerm: string): Promise<Customer[]> => {
  const response = await api.get<Customer[]>(`${CUSTOMERS_URL}/search?lastName=${searchTerm}`)
  return response.data
}

const customerService = {
  getAllCustomers,
  getCustomerById,
  createCustomer,
  updateCustomer,
  deleteCustomer,
  searchCustomers,
}

export default customerService

